### PR TITLE
mds: async dir operation support

### PR DESCRIPTION
**Pull request by ukernel**
Change the order of taking locks to:

1. Lock objects in fs tree first, then lock objects in mdsdir
2. Lock objects according to their depth, in ascending order.
3. Lock objects according to inodes' numbers or dentries' parent
   inode numbers, in ascending order.

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by ukernel** [comment]
kernel patches https://github.com/ceph/ceph-client/tree/ceph-async-unlink

**Comment by jtlayton** [comment]
I'm a bit confused about how this works with hardlinks, and I don't know the MDS code well enough to know how this is avoided:

Suppose the MDS hands out Fx and Du (DIR_UNLINK) caps on a directory. We have a dentry in that directory that points to a file and get back a link count of 1 and it's marked as a PRIMARY_DENTRY.
Later, a different client does a lookup by inode number (maybe via NFS server) on that file, and uses ceph_ll_link() to hardlink it into a different directory. I assume that we will need to revoke Fx or Du on the original parent dir at that point.

Will that occur?

**Comment by ukernel** [comment]
Current code does not handle link count and lookup properly.  Even a client hold Fx and dir unlink cap on a directory, other client still can hardlink file in that directory into other directory.  The hard link issue is somewhat related to lookup issue (https://tracker.ceph.com/issues/39098 comment 25/28). 

I'm thinking about how to implement exclusive lease. Once we have exclusive lease for dentry. For async unlink case, client can hold  Fx and dir unlink cap on directory, exclusive lease on the dentry to unlink, Lx cap on the file to unlink.  This should solve all issue we have.

File: src/include/ceph_fs.h [hunk 1]
```diff
@@ -790,6 +790,13 @@ int ceph_flags_to_mode(int flags);
 #define CEPH_CAP_LOCKS (CEPH_LOCK_IFILE | CEPH_LOCK_IAUTH | CEPH_LOCK_ILINK | \
 			CEPH_LOCK_IXATTR)
 
+/* cap masks async dir operations */
+#define CEPH_CAP_DIR_CREATE    CEPH_CAP_FILE_CACHE
+#define CEPH_CAP_DIR_UNLINK    CEPH_CAP_FILE_RD
```
Comment (jtlayton) [inline src/include/ceph_fs.h]: So if I'm given CEPH_CAP_DIR_UNLINK on a directory, what else do I need in order to unlink a dentry in that directory? I'll also need a valid dentry for the target, but the kernel client already does a lookup on it beforehand, so we can be reasonably sure that the dentry is present.

That seems like a workable model, and I Iike the basic lock cache concept.

What I'd really like to see is some sort of written spec for this that client authors can follow. It doesn't have to be anything formal, but I think we need to clearly document what each of the above caps is actually granting the client, and what else (if anything) it needs to acquire before performing these operations.

Also, are there ever situations where we have DIR_CREATE but not DIR_UNLINK? How much different is the mds locking for those cases?

Comment (ukernel) [reply src/include/ceph_fs.h]: DIR_CREATE and DIR_UNLINK are separate from each other. Locking for create requires policy lock (for subtree file layout),  unlink does not need that lock.

**Comment by jtlayton** [comment]
I've been forward-porting the companion kernel patchset, and I still don't quite understand how this will work. Your changes to my earlier set had this:
```
+        spin_lock(&dentry->d_lock);
+        di = ceph_dentry(dentry);
+       /* - We are holding CEPH_CAP_FILE_EXCL, which implies
+        * CEPH_CAP_FILE_SHARED.
+        * - Only support async unlink for primary linkage */
+       if (atomic_read(&ci->i_shared_gen) != di->lease_shared_gen ||
+           !(di->flags & CEPH_DENTRY_PRIMARY_LINK))
+               ret = 0;
+        spin_unlock(&dentry->d_lock);
```
I don't quite understand why checking for PRIMARY_LINK is enough. With this set, does the MDS avoid handing out CEPH_CAP_DIR_UNLINK if any dentry inode below that directory happens to be hardlinked?

**Comment by ukernel** [comment]
hard link in directory does not affect if mds issues CEPH_CAP_DIR_UNLINK.  async unlink remote dentry  is not supported because locking order of unlink remote dentry can be different from locking order of unlink primary dentry. lock cache can not handle it

**Comment by jtlayton** [comment]
Zheng, do you have a more recent branch with this code? I was wanting to do some testing with the corresponding kernel patches, but this doesn't seem to apply to current master branch.

**Comment by ukernel** [comment]
new code is ready for review

**Comment by jtlayton** [comment]
I merged this into this morning's master branch and ran xfstests against it with a kernel based on the ceph/testing branch (no async unlink or create support).

It looks like this has broken setgid inheritance from directories (generic/314 test in xfstests).

Basically, make a directory and set the mode to 02755 and then make a new directory inside it. The resulting directory doesn't should have the setgid bit set, but when running against a cluster with this, it doesn't.

It is possible that this is something recently broken in master and not due to this set. I'll test that soon to make sure.

**Comment by jtlayton** [comment]
Confirmed. I don't see this on master, so I think this is a regression in this set.

File: src/mds/MDCache.cc [hunk 2]
```diff
@@ -8286,24 +8334,48 @@ int MDCache::path_traverse(MDRequestRef& mdr, MDSContextFactory& cf,     // who
...
 	    dout(20) << " added null " << *dn << dendl;
+
+	    if (rdlock_path) {
+	      lov.clear();
+	      if (xlock_dentry) {
+		if (depth > 0 || !mdr->lock_cache) {
+		  lov.add_wrlock(&cur->filelock);
+		  lov.add_wrlock(&cur->nestlock);
+		  if (rdlock_authlock)
+		    lov.add_rdlock(&cur->nestlock);
```
Comment (ukernel) [inline src/mds/MDCache.cc]: xfstest failure was caused by typo here

Comment (jtlayton) [reply src/mds/MDCache.cc]: Confirmed. The new set seems to fix it.

**Comment by jtlayton** [comment]
Nice work, @ukernel !

Initial testing looks good. I ran xfstests on a kernel with the experimental async unlink patches vs. a vstart cluster with this series and it worked just fine. I also confirmed that async unlink is working. A few of the tests seemed to run a little faster than before, but that could just be a fluke.

**Comment by jtlayton** [comment]
I started looking over the series in earnest, but it's rather large. I wonder if we ought to break this up a little and merge it in pieces. Start with the earlier cleanup patches and then move to merge the rest of the series afterward?

It'd be nice to merge the patches that aren't expected to cause any behavior change first so we can do a full QA run and make sure we don't see any regressions.

Also, this series currently fails a number of tests in `./bin/ceph_test_libcephfs`, mostly dealing with delegations and hardlinks.

```
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] LibCephFS.HardlinkNoOriginal
[  FAILED  ] LibCephFS.Nlink
[  FAILED  ] LibCephFS.DelegMultiClient
[  FAILED  ] LibCephFS.DelegSingleClient
```

**Comment by ukernel** [comment]
ceph_test_libcephfs failures are resolved by "mds: handle link request with zero depth filepath2 "

**Comment by jtlayton** [comment]
Yes, push from earlier today seems to have fixed the gtest regressions, and xfstests (still running) looks good so far.

I started trying to review the patches but this set is **huge** and touches a lot of code. A lot of the earlier patches in the series seem to be cleanup and preparation patches that shouldn't change behavior much. I think it'd be good to merge this in stages, starting with those.

Once the cleanup patches are in, we can do a full QA cycle and focus on reviewing the ones that make material changes. Thoughts?

**Comment by ukernel** [comment]
preparation and cleanup are at https://github.com/ceph/ceph/pull/30972

File: src/mds/Locker.cc [hunk 3]
```diff
@@ -2234,6 +2285,11 @@ int Locker::issue_caps(CInode *in, Capability *only_cap)
 
     // add in any xlocker-only caps (for locks this client is the xlocker for)
     allowed |= xlocker_allowed & in->get_xlocker_mask(it->first);
+    if (in->is_dir()) {
+      allowed &= ~CEPH_CAP_ANY_DIR_OPS;
+      if (allowed & CEPH_CAP_FILE_EXCL)
+	allowed |= cap->get_lock_cache_allowed();
```
Comment (jtlayton) [inline src/mds/Locker.cc]: nit: funky indentation here

Comment (ukernel) [reply src/mds/Locker.cc]: indentation is correct here. the 'allowed ...' line is a tab, the 'if (allowed ...' line is 6 spaces.

**Comment by ukernel** [comment]
the main reason is not implement code that issue DIR_FOO caps to single client. reuse the code that mds only issues exclusive cap to single client.

If client only have DIR_FOO for a directory, another clients also can modify the directory.  we don't have exclusive lease,  so multiple client can modify a dentry at the same time.   Lease can timeout, it's less reliable than caps,   this make exclusive lease hard to implement.

File: src/mds/Locker.cc (hunk 3 above)
Comment (jtlayton) [reply src/mds/Locker.cc]: Ok, thanks. github--

**Comment by jtlayton** [comment]
Another question too. If I do an asynchronous create, does the MDS assume that I have full caps for the new inode? The way I would envision creates working is that we do them asynchronously and send the new inode number, and just have the client assume that it has all caps for the new inode. Once the create reply is returned (successfully), the MDS could then just revoke those caps at any time to force the client to flush everything back.

**Comment by ukernel** [comment]
On Fri, Nov 15, 2019 at 1:02 AM jtlayton <notifications@github.com> wrote:
>
> @jtlayton commented on this pull request.
>
> One reason is not implement code that issue DIR_FOO caps to single client. reuse the code that mds only issues exclusive cap to single client.
>
> Another reason is: If client only have DIR_FOO for a directory, another clients also can modify the directory. We don't have exclusive lease, so multiple client can modify a dentry at the same time. Lease can timeout, it's less reliable than caps, this make exclusive lease hard to implement.
>
> My thinking was that having the client manipulate two caps for this is somewhat more complex than is necessary for the protocol. To wit: If I have DIR_UNLINK or DIR_CREATE and don't have FILE_EXCLUSIVE, does that grant me anything?
>

Just indicate mds can use lock cache for given type of operation (use
less CPU cycles to handle the operation).

**Comment by ukernel** [comment]
On Fri, Nov 15, 2019 at 5:56 AM jtlayton <notifications@github.com> wrote:

> Another question too. If I do an asynchronous create, does the MDS assume
> that I have full caps for the new inode? The way I would envision creates
> working is that we do them asynchronously and send the new inode number,
> and just have the client assume that it has all caps for the new inode.
> Once the create reply is returned (successfully), the MDS could then just
> revoke those caps at any time to force the client to flush everything back.
>
>
mds does not issue full caps in implementation. But easy to change, just
change all locks' states to EXCL when creating new inode

**Comment by jtlayton** [comment]
> On Fri, Nov 15, 2019 at 5:56 AM jtlayton ***@***.***
> mds does not issue full caps in current implementation. But easy to change, just change all locks' states to EXCL when creating new inode

Ok, good. I think we will need to do that as otherwise nothing grants the client the ability to modify the file before the create reply comes back and it gets the caps for the inode. If we can assume that the client holds full exclusive caps on the created inode when it does a create in a directory where it holds Fx and DIR_CREATE (Dc), then we can just rely on the normal cap revoke mechanism to handle the rest.

**Comment by batrick** [comment]
Tests running in 3 hours: http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20191204.231200

**Comment by ukernel** [comment]
hang tests should be fixed by the last commit.  I also killed tests for wip-pdonnell-testing-20191204.231200

**Comment by batrick** [comment]
ACK. As discussed during standup, please just do your own integration testing on this so you aren't blocked waiting for me. I think `--subset 0/2` on all the suites should be good.

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2019-12-10_08:41:40-fs-wip-mds-locking-order-testing-basic-smithi/

**Comment by ukernel** [comment]
An old run http://pulpito.ceph.com/zyan-2019-12-06_02:10:01-fs-wip-mds-locking-order-testing-basic-smithi/.

The failed snaptests are because lic of rhel7 has no syncfs syscall

File: src/mds/SimpleLock.h [hunk 4]
```diff
@@ -142,10 +142,9 @@ class SimpleLock {
       case CEPH_LOCK_INEST: return "inest";
       case CEPH_LOCK_IXATTR: return "ixattr";
       case CEPH_LOCK_ISNAP: return "isnap";
-      case CEPH_LOCK_INO: return "ino";
       case CEPH_LOCK_IFLOCK: return "iflock";
       case CEPH_LOCK_IPOLICY: return "ipolicy";
-      default: ceph_abort(); return std::string_view();
+      default: return "unkown";
```
Comment (fullerdj) [inline src/mds/SimpleLock.h]: nit: "unknown"

File: src/mds/Locker.cc [hunk 5]
```diff
@@ -1741,6 +1741,22 @@ void Locker::xlock_import(SimpleLock *lock)
   lock->get_parent()->auth_pin(lock);
 }
 
+void Locker::xlock_downgrade(SimpleLock *lock, MutationImpl *mut)
```
Comment (fullerdj) [inline src/mds/Locker.cc]: nit: "instrcuts" => "instructs" in commit message

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2019-12-10_08:41:40-fs-wip-mds-locking-order-testing-basic-smithi/

**Comment by jtlayton** [comment]
> The failed snaptests are because libc of rhel7 has no syncfs syscall. No clue why libcephfs interface failed once

Are you sure? syncfs has been around a long time. In fact, looking at my centos7.7 VM:
```
/usr/include/unistd.h:extern int syncfs (int __fd) __THROW;
```

**Comment by ukernel** [comment]
you are right. I wrongly read the log. snaptests failures were caused by a bug in this PR

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2019-12-18_13:40:19-fs-wip-mds-locking-order-testing-basic-smithi/

I think this is ready to merge

**Pull request merged**